File: haskell/app/Day11.hs
-- ~/~ begin <<docs/day11.md#haskell/app/Day11.hs>>[init]
module Main where

import qualified Data.Text as T
import qualified Data.Text.IO as T.IO
import Control.Monad.IO.Class (MonadIO, liftIO)
import Data.MemoTrie (memoFix)

readInput :: (MonadIO m) => m [Int]
readInput = liftIO $ fmap (read . T.unpack) . T.words <$> T.IO.getContents

ndigits :: Int -> Int
ndigits x
    | x < 10 = 1
    | otherwise = ndigits (x `div` 10) + 1

countStones :: ((Int, Int) -> Int) -> (Int, Int) -> Int
countStones f (x, n)
    | n == 0            = 1
    | x == 0            = f (1, n - 1)
    | even $ ndigits x  = f (x `div` q, n - 1) + f (x `mod` q, n - 1)
    | otherwise         = f (x * 2024, n - 1)
    where q = 10^(ndigits x `div` 2)

blink :: [Int] -> Int -> Int
blink input n = sum [cs (i, n) | i <- input]
    where cs = memoFix countStones

main :: IO()
main = do
    input <- readInput
    print $ blink input 25
    print $ blink input 75
-- ~/~ end

File: src/Day11.jl
# ~/~ begin <<docs/day11.md#src/Day11.jl>>[init]
module Day11

using Memoize: @memoize

@memoize function count_stones(x::Int, n::Int)
    if n == 0
        1
    elseif x == 0
        count_stones(1, n - 1)
    elseif ndigits(x) % 2 == 0
        f = 10^(ndigits(x) ÷ 2)
        count_stones(x ÷ f, n - 1) + count_stones(x % f, n - 1)
    else
        count_stones(x * 2024, n - 1)
    end
end

function main(io::IO)
    input = read(io, String) |> split .|> (x->parse(Int, x))
    part1 = sum(count_stones(x, 25) for x in input)
    part2 = sum(count_stones(x, 75) for x in input)
    return part1, part2
end

end
# ~/~ end
